pallets/auctions/src/lib.rs

```diff
@@ -160,171 +154,279 @@ pub mod pallet {
+			general_data.token.0,
+			general_data.token.1,
 		)?;
-		<AuctionOwnerById<T>>::remove(id);
-		<Auctions<T>>::remove(id);
+
+		<AuctionOwnerById<T>>::remove(auction_id);
+		<Auctions<T>>::remove(auction_id);
+
+		Self::deposit_event(Event::AuctionRemoved(auction_id));
+
 		Ok(())
 	}
 
-	fn bid(bidder: Self::AccountId, id: Self::AuctionId, value: Self::Balance) -> DispatchResult {
-		<Auctions<T>>::try_mutate_exists(id, |auction| -> DispatchResult {
-			// Basic checks before a bid can be made
-			let mut auction = auction.as_mut().ok_or(Error::<T>::AuctionNotExist)?;
-			let block_number = <frame_system::Pallet<T>>::block_number();
-			ensure!(bidder != auction.owner, Error::<T>::BidOnOwnAuction);
-			ensure!(block_number > auction.start, Error::<T>::AuctionNotStarted);
-			ensure!(block_number < auction.end, Error::<T>::AuctionAlreadyConcluded);
-			ensure!(value >= auction.minimal_bid, Error::<T>::InvalidBidPrice);
-			if let Some(ref current_bid) = auction.last_bid {
-				ensure!(value > current_bid.1, Error::<T>::InvalidBidPrice);
-				// Unlock funds from the previous bid
-				<T as crate::Config>::Currency::remove_lock(AUCTION_LOCK_ID, &current_bid.0);
-			} else {
-				ensure!(!value.is_zero(), Error::<T>::InvalidBidPrice);
-			}
-			// Lock funds
-			<T as crate::Config>::Currency::set_lock(AUCTION_LOCK_ID, &bidder, value, WithdrawReasons::all());
-			auction.last_bid = Some((bidder, value));
-			// Set next minimal bid
-			let minimal_bid_step = Permill::from_percent(BID_STEP_PERC).mul_floor(value);
-			auction.minimal_bid = value.checked_add(&minimal_bid_step).ok_or(Error::<T>::BidOverflow)?;
-			// Avoid auction sniping
-			let time_left = auction
-				.end
-				.checked_sub(&block_number)
-				.ok_or(Error::<T>::TimeUnderflow)?;
-			if time_left < BID_ADD_BLOCKS.into() {
-				auction.end = block_number + BID_ADD_BLOCKS.into();
+	fn validate_bid(
+		bidder: &<T>::AccountId,
+		general_auction_data: &GeneralAuctionData<T>,
+		value: BalanceOf<T>,
+	) -> DispatchResult {
+		let block_number = <frame_system::Pallet<T>>::block_number();
+		ensure!(bidder != &general_auction_data.owner, Error::<T>::BidOnOwnAuction);
+		ensure!(block_number > general_auction_data.start, Error::<T>::AuctionNotStarted);
+		ensure!(
+			block_number < general_auction_data.end,
+			Error::<T>::AuctionAlreadyConcluded
+		);
+		ensure!(value >= general_auction_data.next_bid_min, Error::<T>::InvalidBidPrice);
+
+		if let Some(ref current_bid) = general_auction_data.last_bid {
+			ensure!(value > current_bid.1, Error::<T>::InvalidBidPrice);
+		} else {
+			ensure!(!value.is_zero(), Error::<T>::InvalidBidPrice);
+		}
+
+		Ok(())
+	}
+
+	fn close_auctions(current_block: T::BlockNumber) -> DispatchResult {
+		for (auction_id, _) in <AuctionEndTime<T>>::drain_prefix(&current_block) {
+			let auction = <Auctions<T>>::get(auction_id).ok_or(Error::<T>::AuctionNotExist)?;
+
+			match &auction {
+				Auction::English(auction_object) => {
+					auction_object.close()?;
```

Comment:
I think `close_auctions ` will fail when closing any first auction fails, potentially leaving some of auction open even if they should be closed ? 

I guess that is not desired behavior.l

Reply:
correct, issue should disappear after moving close to extrinsic (next on todo list)